File: AppSets-Servers-Spring/Share/src/main/kotlin/xcj/app/util/StaticProvider.kt
package xcj.app.util

class StaticProvider<T> {

    private var t: T? = null

    val current: T
        get() = t ?: throw IllegalStateException("current is null, nothing can provide!")

    fun provider(provider: () -> T?) {
        t = provider()
    }

    infix fun provide(t: T?) {
        provider { t }
    }

    infix fun provide(provider: () -> T?) {
        provider(provider)
    }
}

fun <T> staticProvider(provider: () -> T?): StaticProvider<T> {
    return StaticProvider<T>().apply {
        provide(provider)
    }
}

fun <T> staticProvider(): StaticProvider<T> {
    return StaticProvider()
}

File: AppSets-iOS/AppSets/com/app/appsets/util/lang/StaticProvider.swift
//
//  StaticProvider.swift
//  AppSets
//
//  Created by caiju Xu on 2024/7/4.
//

import Foundation


class StaticProvider<T> {
    
    private var key:String? = nil
    
    var t:T? = nil
    
    var current: T {
        get {
            if t == nil {
                Exception.IllegalStateException(message: "current is null, nothing can provide!")
            }
            return t!
            
        }
    }
    
    func provide(provider:()->T?) {
        t = provider()
    }
    
    func provide(t:T?) {
        provide(provider: { t })
    }
    
    
}

func staticProvider<T>(_ t:T? = nil) -> StaticProvider<T> {
    let staticProvider = StaticProvider<T>()
    staticProvider.provide(t: t)
    return staticProvider
}


func staticProvider<T>(_ provider:()->T?) -> StaticProvider<T> {
    let staticProvider = StaticProvider<T>()
    staticProvider.provide(provider:provider)
    return staticProvider
}


